File: go/419.go
package main

func traverse(r int, c int, board [][]byte) {
	if r < 0 || c < 0 || r >= len(board) || c >= len(board[r]) || board[r][c] == '.' {
		return
	}

	board[r][c] = '.'
	traverse(r-1, c, board)
	traverse(r+1, c, board)
	traverse(r, c-1, board)
	traverse(r, c+1, board)

}

func countBattleships(board [][]byte) int {
	result := 0

	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[i]); j++ {
			if board[i][j] == 'X' {
				result++
				traverse(i, j, board)
			}
		}
	}

	return result
}

File: zig/419.zig
const std = @import("std");

fn solution(board: [][]u8) u32 {
    var result: u32 = 0;

    for (0..board.len) |i| {
        for (0..board[i].len) |j| {
            if (board[i][j] == 'X') {
                const top = i > 0 and board[i - 1][j] == 'X';
                const left = j > 0 and board[i][j - 1] == 'X';
                if (!top and !left) result += 1;
            }
        }
    }

    return result;
}

const expect = std.testing.expect;

test "test case #1" {
    var row0 = [_]u8{ 'X', '.', '.', 'X' };
    var row1 = [_]u8{ '.', '.', '.', 'X' };
    var row2 = [_]u8{ '.', '.', '.', 'X' };
    var board = [_][]u8{ &row0, &row1, &row2 };

    try expect(solution(&board) == 2);
}

test "test case #2" {
    var row0 = [_]u8{'.'};
    var board = [_][]u8{&row0};

    try expect(solution(&board) == 0);
}
